jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java

```diff
@@ -394,15 +402,20 @@ private boolean tryRetain()
         {
             return _cache.computeIfPresent(_cacheKey, (s, cachingHttpContent) ->
             {
-                _buffer.retain();
+                RetainableByteBuffer buffer = _buffer.get();
+                if (buffer == null)
+                    return null;
+                buffer.retain();
                 return cachingHttpContent;
             }) != null;
         }
 
         @Override
         public void release()
         {
-            _buffer.release();
+            RetainableByteBuffer buffer = _buffer.get();
+            if (buffer != null && buffer.release())
+                _buffer.set(null);
```

Comment:
Use `AtomicReference.getAndUpdate()` instead.


```suggestion
            _buffer.getAndUpdate(buffer -> 
            {
                if (buffer != null && buffer.release())
                    return null;
                return buffer;  
            });
```

Reply:
If the `_buffer` field can ever be null (see my previous comment), I don't mind this suggestion. But my understanding of the code is that `_buffer` can never be null.

Reply:
@lorban, we are deliberately nulling the reference so that if the buffer is released this content still works if writeTo is called.   I'm not exactly sure of the use-case for that, but we do have a test for it.